cli/azd/pkg/tools/sqlcmd/sqlcmd.go

```diff
@@ -0,0 +1,376 @@
+}
+
+func extractFromTar(src, dst string) (string, error) {
+	bz2File, err := os.Open(src)
+	if err != nil {
+		return "", err
+	}
+	defer bz2File.Close()
+
+	bz2Reader := bzip2.NewReader(bz2File)
+
+	var extractedAt string
+	// tarReader doesn't need to be closed as it is closed by the gz reader
+	tarReader := tar.NewReader(bz2Reader)
+	for {
+		fileHeader, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			return extractedAt, fmt.Errorf("did not find sqlcmd cli within tar file")
+		}
+		if fileHeader == nil {
+			continue
+		}
+		if err != nil {
+			return extractedAt, err
+		}
+		// Tha name contains the path, remove it
+		fileNameParts := strings.Split(fileHeader.Name, "/")
+		fileName := fileNameParts[len(fileNameParts)-1]
+		// cspell: disable-next-line `Typeflag` is comming fron *tar.Header
+		if fileHeader.Typeflag == tar.TypeReg && fileName == "sqlcmd" {
+			filePath := filepath.Join(dst, fileName)
+			sqlCmdCliFile, err := os.OpenFile(filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, os.FileMode(fileHeader.Mode))
+			if err != nil {
+				return extractedAt, err
+			}
+			defer sqlCmdCliFile.Close()
+			/* #nosec G110 - decompression bomb false positive */
+			_, err = io.Copy(sqlCmdCliFile, tarReader)
+			if err != nil {
+				return extractedAt, err
+			}
+			extractedAt = filePath
+			break
+		}
+	}
+	if extractedAt != "" {
+		return extractedAt, nil
+	}
+	return extractedAt, fmt.Errorf("extract from tar error. Extraction ended in unexpected state.")
+}
+
+// extractSqlCmdCli gets the sqlCmd cli from either a zip or a tar.gz
+func extractSqlCmdCli(src, dst string) (string, error) {
+	if strings.HasSuffix(src, ".zip") {
+		return extractFromZip(src, dst)
+	} else if strings.HasSuffix(src, ".tar.bz2") {
+		return extractFromTar(src, dst)
+	}
+	return "", fmt.Errorf("Unknown format while trying to extract")
+}
+
+// getSqlCmdCliImplementation defines the contract function to acquire the sqlCmd cli.
+// The `outputPath` is the destination where the sqlCmd cli is place it.
+type getSqlCmdCliImplementation func(
+	ctx context.Context,
+	transporter policy.Transporter,
+	sqlCmdVersion semver.Version,
+	extractImplementation extractSqlCmdCliFromFileImplementation,
+	outputPath string) error
+
+// extractSqlCmdCliFromFileImplementation defines how the cli is extracted
+type extractSqlCmdCliFromFileImplementation func(src, dst string) (string, error)
+
+// downloadSqlCmd downloads a given version of sqlCmd cli from the release site.
+func downloadSqlCmd(
+	ctx context.Context,
+	transporter policy.Transporter,
+	sqlCmdVersion semver.Version,
+	extractImplementation extractSqlCmdCliFromFileImplementation,
+	path string) error {
+
+	binaryName := func(platform string) string {
+		return fmt.Sprintf("sqlcmd-%s", platform)
+	}
+
+	systemArch := runtime.GOARCH
+	// arm and x86 not supported (similar to bicep)
+	var releaseName string
+	switch runtime.GOOS {
+	case "windows":
```

Comment:
All of this custom download/install logic is starting to scare me... `azd` is not a package manager but instead we should consider leveraging existing package manager like winget or choco to help install missing required components.  

I don't think maintaining install/download scripts for various tools is something we want to get into the business of maintaining.

Reply:
Talked ofline to Wallace about this.
Maybe we can keep bringing stand-alone binaries to ~/.azd folder if the tool is under Microsoft umbrella. Like gh-cli, bicep-cli, swa-cli, and now sqlcmd-cli.

We definitely don't want to use azd to pull other tools like terraform or something else.

The main idea of using standalone binaries is to have only azd to know about it and be the less invasive to customer's PC.  We also don't want to deal with min versions and updates when asking for tools to be installed in the system.